modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -278,6 +278,7 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
                         if not batch_accumulator.inform_batch(data, sample_ids, target, weights):
                             stopw.start("FetchBatch", resume=True)
                             stopw.start("IndivFetchBatch", overwrite=True)
+                            self._num_samples += self._batch_size
```

Comment:
When we accumulate batches in `batch-then-sample`, we forget to increment `self._num_samples`.

Reply:
I actually do not think this logic is correct. Num samples tracks how many samples we trained on, not how many samples in total we passed. In StB, it would be incremented just as often as the downsampled set is big. In BtS, we with this change do it as often as we have pre downsampling items. The intention is to track the number of items we actually used for model updates. Hence I dont think they change is correct, but at least not consistent.

Reply:
Oh, good that we encountered this! I always assumed `num_samples` would be the number of samples **considered** for training in the sense of training interval size. I think, it makes sense to report the number of samples in the training interval, the number of samples we trained on, and potentially also the number of samples we passed. I can be valuable to have all of these metrics available in the supervisor.

Reply:
@MaxiBoether I understand your comment. But we *did* use this number `self._num_samples` previously to track how many samples in total we passed. There are two points to prove this:

1. Before having this batch accumulation logic, we incremented `self._num_samples` once per batch, by the pre-downsampled batch size. https://github.com/eth-easl/modyn/blob/988bd013db897e7f15c40865539e732f62c33d84/modyn/trainer_server/internal/trainer/pytorch_trainer.py#L326
2. In client log tracker, we compare "the number of samples passed" with the "number of samples per epoch" to determine the current epoch number. https://github.com/eth-easl/modyn/blob/988bd013db897e7f15c40865539e732f62c33d84/modynclient/client/internal/utils/training_status_tracker.py#L103 I confirmed that the first number stems from the `self._num_samples`, and the second number stems from the data size after (possible) presampling at selector. https://github.com/eth-easl/modyn/blob/988bd013db897e7f15c40865539e732f62c33d84/modyn/supervisor/internal/pipeline_executor/pipeline_executor.py#L661

Reply:
This could for sure explain the epoch number inconsistencies.  I think we should synchronize this somehow. Should we

1) track `_ num_samples_seen` and `_num_samples_trained` in the logs

and

2) report `_num_samples_trained` to the client such that the user is not confused why different sampling strategies return the same number of samples in the progress bar? I think that would be the cleanest fix here. Or do you disagree?

Reply:
I get your idea, but if we only report `_num_samples_trained` to the client, then the client has no way to determine the current epoch number. The `_ num_samples_seen` and `_num_samples_trained` only differ by the ratio. I suppose only reporting one of them is enough?

Reply:
@MaxiBoether Probably there are only two, I was not fully sure what you meant with `how many samples in total we passed` exactly. If this is the total number of samples in the training interval set by the supervisor, then there are only two metrics!

@XianzheMa Generally, I think, the progress bar reporting is not extremely critical and should just give a rough estimation of the progress. I would argue that the exact number isn't too critical there. It's more important to have the different metrics available in the supervisor / the supervisor log file.

Reply:
Hmh, I am a bit confused. Generally, yes, complete logs are the most important thing. With the current implementation (for me at least), the warmup triggers show twice the number of epochs. As long as the epoch count is correct and the relative progression of the progress bar makes sense, I guess reporting either is ok. Having the correct absolute number ( how many samples have we trained on / how many samples will we train on with train in terms of backward pass) would be best but if it's a bigger problem let's go for the easier solution

Reply:
Okay. So I will record both `_ num_samples_seen`, `_num_samples_trained` in the logs, and report the former to the client as the current solution!